Creating multidimensional array gives c error:  'Array_fixed__v_Array_fixed_f32_4_1' undeclared

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 5c65e58.f1333be
OS: windows, Microsoft Windows 11 Pro v22631 64-bit
Processor: 24 cpus, 64bit, little endian, 

getwd: D:\vbug
vexe: C:\v\v.exe
vexe mtime: 2024-11-17 16:24:55

vroot: OK, value: C:\v
VMODULES: OK, value: C:\Users\info\.vmodules
VTMP: OK, value: C:\Users\info\Documents\MobaXterm\slash\mx86_64b\var\log\xwin\v_0

Git version: git version 2.42.0.windows.2
Git vroot status: weekly.2024.46-52-gf1333be6-dirty
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


emcc version: Error: 'emcc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 b425ac82
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg src/main.v`

```v
module main

import math.vec

type Vec4 = vec.Vec4[f32]

fn (v Vec4) to_array() [4]f32 {
	return [v.x, v.y, v.z, v.w]!
}

fn main() {
	v := Vec4{0, 0, 0, 1}
	_ := [
		v.to_array()
	]!	
}
```

**What did you expect to see?**

no c error

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: C:/Users/info/Documents/MobaXterm/slash/mx86_64b/var/log/xwin/v_0/main.01JCXFWWTBP485KF3BHW85SATY.tmp.c:7697: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/info/Documents/MobaXterm/slash/mx86_64b/var/log/xwin/v_0/main.01JCXFWWTBP485KF3BHW85SATY.tmp.c:17493: error: 'Array_fixed__v_Array_fixed_f32_4_1' undeclared
... (the original output was 3 lines long, and was truncated to 2 lines)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzNhMWNhMWU3Y2M1YTc4NTQyZDhjYjAiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.ijW_FWMmDwH7hLbZ8TH-R4VlhoSPqdCvi2oWAcSbDEE">Huly&reg;: <b>V_0.6-21330</b></a></sub>

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -1548,15 +1548,16 @@ pub:
 	has_init      bool
 	has_index     bool // true if temp variable index is used
 pub mut:
-	exprs      []Expr // `[expr, expr]` or `[expr]Type{}` for fixed array
-	len_expr   Expr   // len: expr
-	cap_expr   Expr   // cap: expr
-	init_expr  Expr   // init: expr
-	expr_types []Type // [Dog, Cat] // also used for interface_types
-	elem_type  Type   // element type
-	init_type  Type   // init: value type
-	typ        Type   // array type
-	alias_type Type   // alias type
+	exprs        []Expr // `[expr, expr]` or `[expr]Type{}` for fixed array
+	len_expr     Expr   // len: expr
+	cap_expr     Expr   // cap: expr
+	init_expr    Expr   // init: expr
+	expr_types   []Type // [Dog, Cat] // also used for interface_types
+	elem_type    Type   // element type
+	init_type    Type   // init: value type
+	typ          Type   // array type
+	alias_type   Type   // alias type
+	has_callexpr bool   // has expr which needs tmp var to initialize it
 }
 
 pub struct ArrayDecompose {

--- a/vlib/v/checker/containers.v
+++ b/vlib/v/checker/containers.v
@@ -158,7 +158,14 @@ fn (mut c Checker) array_init(mut node ast.ArrayInit) ast.Type {
 			}
 		}
 		for i, mut expr in node.exprs {
-			typ := c.check_expr_option_or_result_call(expr, c.expr(mut expr))
+			mut typ := c.check_expr_option_or_result_call(expr, c.expr(mut expr))
+			if expr is ast.CallExpr {
+				ret_sym := c.table.sym(typ)
+				if ret_sym.kind == .array_fixed {
+					typ = c.cast_fixed_array_ret(typ, ret_sym)
+				}
+				node.has_callexpr = true
+			}
 			if typ == ast.void_type {
 				c.error('invalid void array element type', expr.pos())
 			}

--- a/vlib/v/gen/c/array.v
+++ b/vlib/v/gen/c/array.v
@@ -155,7 +155,8 @@ fn (mut g Gen) fixed_array_init(node ast.ArrayInit, array_type Type, var_name st
 				elem_info := elem_sym.array_fixed_info()
 				g.fixed_array_var_init(g.expr_string(expr), expr.is_auto_deref_var(),
 					elem_info.elem_type, elem_info.size)
-			} else if expr is ast.CallExpr {
+			} else if elem_sym.kind == .array_fixed && expr is ast.CallExpr
+				&& g.table.final_sym(expr.return_type).kind == .array_fixed {
 				elem_info := elem_sym.array_fixed_info()
 				tmp_var := g.expr_with_var(expr, node.expr_types[i], node.expr_types[i])
 				g.fixed_array_var_init(tmp_var, false, elem_info.elem_type, elem_info.size)

--- a/vlib/v/gen/c/array.v
+++ b/vlib/v/gen/c/array.v
@@ -155,6 +155,10 @@ fn (mut g Gen) fixed_array_init(node ast.ArrayInit, array_type Type, var_name st
 				elem_info := elem_sym.array_fixed_info()
 				g.fixed_array_var_init(g.expr_string(expr), expr.is_auto_deref_var(),
 					elem_info.elem_type, elem_info.size)
+			} else if expr is ast.CallExpr {
+				elem_info := elem_sym.array_fixed_info()
+				tmp_var := g.expr_with_var(expr, node.expr_types[i], node.expr_types[i])
+				g.fixed_array_var_init(tmp_var, false, elem_info.elem_type, elem_info.size)
 			} else {
 				if expr.is_auto_deref_var() {
 					g.write('*')

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -697,7 +697,7 @@ fn (mut g Gen) struct_init_field(sfield ast.StructInitField, language ast.Langua
 						tmp_var := g.expr_with_var(sfield.expr, sfield.typ, sfield.expected_type)
 						g.fixed_array_var_init(tmp_var, false, field_unwrap_sym.info.elem_type,
 							field_unwrap_sym.info.size)
-					} else if sfield.expr.exprs.len > 0 && sfield.expr.exprs.any(it is ast.CallExpr) {
+					} else if sfield.expr.has_callexpr {
 						tmp_var := g.expr_with_fixed_array(sfield.expr, sfield.typ, sfield.expected_type)
 						g.fixed_array_var_init(tmp_var, false, field_unwrap_sym.info.elem_type,
 							field_unwrap_sym.info.size)

--- a/vlib/v/tests/assign/array_fixed_init_with_call_test.v
+++ b/vlib/v/tests/assign/array_fixed_init_with_call_test.v
@@ -1,0 +1,23 @@
+module main
+
+import math.vec
+
+type Vec4 = vec.Vec4[f32]
+
+fn (v Vec4) to_array() [4]f32 {
+	return [v.x, v.y, v.z, v.w]!
+}
+
+fn test_main() {
+	v := Vec4{1, 2, 3, 4}
+	_ := [
+		v.to_array(),
+	]!
+	u := [
+		v.to_array(),
+	]!
+	assert u[0][0] == f32(1)
+	assert u[0][1] == f32(2)
+	assert u[0][2] == f32(3)
+	assert u[0][3] == f32(4)
+}

